### DISCUSSION TITLE
Sanitization of $_GET and $_POST variables

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#4824] [General]
The current repository has a mixture of 'sanitized' $_GET/$_POST variables, e.g. `$_GET['order_id'] = (int)$_GET['order_id'];`, that can lead to confusion as we're now tending towards strict comparisons.  When performing those comparisons, I'm finding that it's necessary to track back through a page's processing to see if a $_GET/$_POST has been pre-sanitized.

I'd like to lay the foundation that, going forward, $_GET/$_POST variables are kept in their `(string)` format to reduce the amount of checking required as modules are updated.

Let the discussion begin!

**Comment by scottcwilson** [comment]
I am open to this.

**Comment by torvista** [comment]
I assume the reason for this is just for consistency: assume a GET is always a dirty string until sanitized to another variable?

**Reply by lat9** [reply]
Yes, keeping the $_GET variables as either their original, dirty strings or sanitizing them to be cleansed strings.

**Comment by scottcwilson** [comment]
Clarification: the word "sanitized" might be confusing here because the sanitization (stripping out embedded scripting for example) has already occurred; this is really casting.

**Reply by lat9** [reply]
Although some of the 'sanitization' could be, for example, the use of `ctype_digit` for variables that _should_ be integer values and replacing invalid string values with `'0'` instead of `0`.

**Comment by drbyte** [comment]
The general ethos about this in all core code efforts in the last 5-8 years has been to cast (or clean to a sane value and then cast) to the expected type as early on in page processing as possible. The reasoning behind that is to ensure no dangerous values can be accidentally used/obtained/accessed by any subsequent code (and most plugin or otherwise-customized code is usually found much later in the code).

So, that said, it makes me wonder how many cases exist of such a problem as this discussion begs to address ... ?

Obviously you've run into a case where this exists in enough situations where it caused you to write about it. I'm interested in hearing more about those ;)

**Reply by lat9** [reply]
I can't recall the in-code challenges, but I was trying to get a discussion documented to address PR #4814 ... to set a stake in the sand for changes going forward.

**Comment by drbyte** [comment]
Reading through this again. And related discussions/issues.

**Comment by drbyte** [comment]
The reason for the pre-sanitization of many GET/POST values was two-fold: making sites less-hackable by rogue actors, and also to comply with PA-DSS and pass penetration-testing. 

We did that because all kinds of unvetted "hacks" were posted on the forum (and osc and other osc-competitors' forums) suggesting how storeowners could alter the functionality of things for their own use ... and many of those didn't have any idea about secure-coding practices, thus resulting in sites getting pwned. 
This extended to various addons as well.

Any custom code that directly accessed any GET/POST var instead of the sanitized variable that "should" be used, was open to vulnerability.

We tried pulling all GET/POST/etc vars into a Registry object and prohibiting any direct access to original GET/POST unless going thru that registry ... but that broke nearly every plugin/contribution/addon across the board.
Then we settled on the current approach used in the Admin, because the Admin is a slightly more controlled environment than the storefront, as it is less-customized, and way less traffic.

Embracing Laravel lets us go back to the Registry idea, because it handles all of this stuff already, without us having to do it bespoke for the ZC framework. Like before, it just requires that all core code use the sanitized values. And by rewriting everything to depend on that will break all unsecure "hacks" that have been posted online. But it will also break almost all plugins/contributions, as well.

So, the problem is circular.

The best time to truly do it the modern way is with a huge rewrite. But a lot of stores won't upgrade to that if they're heavily customized.
It's a perpetual challenge.

So.... 

To lat9's original point of this discussion, I agree, moving foward: 
- core code should always use a var that's already been sanitized, never GET/POST
- such vars should be declared as early as possible in code flow, preferably not sooner than they would be "used", but no later either
- GET/POST should never be trusted: always treat it as dirty

I guess we could incorporate some more helper functions for sanitizing things that might require complex logic, instead of merely casting to int/string/etc.

Now ... how much refactoring of existing code is needed to get rid of direct-use of GET/POST vars?

(I wouldn't get rid of existing sanitization that's been applied to GET/POST already, as we can't count on plugins/contributions/addons doing any refactoring.)

**Comment by scottcwilson** [comment]
I think this is the direction you are going - could we perhaps *use* the sanitization layer to produce an array - a dictionary - of sanitized variables?

**Comment by zcwilt** [comment]
We do at the moment instantiate a Laravel request object in both catalog and store .
This is done in includes/application_laravel

This however only gives you  a class that allows easy access to the request parameters and does nothing for sanitization. 

Laravel addresses sanitization in 2 ways, and at the points that matter the most. 

Firstly, its models and query builder use PDO to bind variables to a query to mitigate SQL injection.

Secondly, blade templates use a mustache system e.g {{ $somevar }} to escape output variables and mitigate XSS attacks etc.

Zen Cart has the Bind Vars system to help mitigate SQL injections, but has no system for template output (would require a compiling template system)

I can't stress more highly the need to use bindvars in sql queries, as even now we have code type casting parameters (e.g. (int)$_GET['product_id'] for queries.) 

We have introduced Laravel Models/Eloquent into v157/v158 but converting code to use that is a huge undertaking although could be done incrementally


As Chris has mentioned when we were going through PADSS we had a choice.
Rewrite all of core code to cleanup core code and use some escaping output methods which I guess at the time was  multiple person years of effort or write a "quick" kludge.

The quick "kludge" is the request sanitizer class

**Comment by scottcwilson** [comment]
> The quick "kludge" is the request sanitizer class

I actually thought it was kind of brilliant.   Maybe a shortcut but certainly not a kludge.
